### ISSUE TITLE
Tracking Issue for slice::array_chunks

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#74985]
The feature gate for the issue is `#![feature(array_chunks)]`.

Also tracks `as_(r)chunks(_mut)` in `#![feature(slice_as_chunks)]` -- these were FCPed starting at https://github.com/rust-lang/rust/issues/74985#issuecomment-2769963395 below.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also uses as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] ~~Restrict `N` to values greater than 0 at compile time instead of using a runtime panic.~~ https://github.com/rust-lang/rust/pull/99471
- [x] Support calls where `N == 0` is unreachable due to conditionals https://github.com/rust-lang/rust/issues/74985#issuecomment-2087722186
  - It was decided that a runtime panic is appropriate: https://github.com/rust-lang/rust/issues/74985#issuecomment-2751861090 
- [x] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])
  - [x] slice_as_chunks: #139656
  - [ ] array_chunks

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

- [ ] Iterator vs slices, https://github.com/rust-lang/rust/issues/76354

### Implementation history

- implemented in #74373
- mut version in #75021
- `as_chunks{_mut}` in #76635
- `as_rchunks{_mut}` and unchecked versions in #78818
- const `as_(r)chunk` in #111453
- https://github.com/rust-lang/rust/pull/139656

**Comment by joshlf** [comment]
Is there any intention of providing the reverse versions of these functions?

**Comment by thomcc** [comment]
> Is there any intention of providing the reverse versions of these functions?

What would that look like?

**Comment by joshlf** [comment]
Same API, but iterating backwards. I have a use case in which I need to convert either a prefix or a suffix of a slice to an array reference. This API is perfect for the prefix use case, and I'd love to be able to use it for the suffix use case as well.

**Comment by newpavlov** [comment]
@joshlf 
Both slice of arrays and custom iterator type would implement the [`DoubleEndedIterator`](https://doc.rust-lang.org/beta/std/iter/trait.DoubleEndedIterator.html), so your use case does not need a separate reverse method.

**Comment by newpavlov** [comment]
Ah, I misunderstood your use case, you want `[0, 1, 2, 3, 4, 5, 6]` to be split into `[4, 5, 6], [1, 2, 3], [0]`, not into `[6], [3, 4, 5], [0, 1, 2]`. It indeed requires a separate method.

**Comment by joshlf** [comment]
Yep, exactly 😃

**Comment by thomcc** [comment]
Yeah, array_rchunks is definitely worth having.

**Comment by sffc** [comment]
Also useful would be the inverse operation: `[[0, 1, 2], [3, 4, 5]]` to `[0, 1, 2, 3, 4, 5]`. (from `&[[T; 3]]` to `&[T]`)

**Comment by ChayimFriedman2** [comment]
Do we need to wait until `const_evaluatable_unchecked` is stabilized so we can deny `N = 0` at compile time to stabilize this feature? Perhaps we can just mention in the documentation that currently it panics but it may be uplifted to a compile time error in the future.

**Comment by fzyzcjy** [comment]
Thanks!

**Comment by sffc** [comment]
Another workaround is to map the slice chunks to array chunks.

```rust
my_slice.chunks(3).map(|s| {
    let mut a = [0; 3];
    a.copy_from_slice(s);
    a
})  // now you have an array_chunks iterator
```

**Comment by clarfonthey** [comment]
One potential use case that might be worth having a dedicated method:

```rust
impl<T> [T] {
    fn get_chunk<const N: usize>(&self, idx: usize) -> &[T; N] {
        &self[idx..idx + N].as_chunks::<N>().0[0]
    }
    fn get_chunk_mut<const N: usize>(&mut self, idx: usize) -> &mut [T; N] {
        &mut self[idx..idx + N].as_chunks_mut::<N>().0[0]
    }
}
```

If no one takes it up, at some point, I might implement this myself.

**Comment by jDomantas** [comment]
> Do we need to wait until const_evaluatable_unchecked is stabilized so we can deny N = 0 at compile time to stabilize this feature? Perhaps we can just mention in the documentation that currently it panics but it may be uplifted to a compile time error in the future.

That is a stricter requirement - valid code then must *prove* that N != 0 in addition to avoiding calling `array_chunks::<0>()`. With your suggestion otherwise valid programs could start failing to compile:

```rust
fn foo<const N: usize>() {
    for chunk in get_slice().array_chunks() {
        //                   ^^^^^^^^^^^^ error: can't prove that N != 0
        do_whatever(chunk);
    }
}

fn main() {
    foo::<3>();
}
```

**Comment by Liamolucko** [comment]
Has a version of `array_chunks` for iterators rather than slices been considered?

**Comment by clarfonthey** [comment]
Most of the iterator adapters that have been slowly added to libstd made their start in [itertools](https://docs.rs/itertools/latest/itertools/), so, you should probably start by looking what that offers, and then after that make a case for why it should be in libstd.

From what I understand, though, a lot of those methods are slowly being added to the `Iterator` trait directly, though.

This isn't really the place for that kind of discussion, though, and it might be best to open a separate issue in the RFCs repo, on Zulip, or elsewhere.

**Comment by mina86** [comment]
Should we also have a TryFrom implementation which converts slice into
chunks and fails if the slice cannot be evenly divided?  I found that
I often need such an operation and it could be more concise if there
was a try_from that I could map_err or unwrap when checking for empty
reminder.  With as_chunks I need to interrupt the flow of the code by
adding an `if !reminder.is_empty() { ... }` check.

https://github.com/rust-lang/rust/pull/105316 is how this would look.

**Comment by joboet** [comment]
With auto traits and negative `impl`s, the check can be made via the trait system, so it works even in `cargo check`:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d2f9c9b356807ba132d4e3e97708bdaf

**Comment by lcnr** [comment]
it can not, because partially applicable negative impls aren't really supported for auto traits rn/and once they are, they won't do what you expect them to do:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c0ed93fccc6cfd16e067e22bc834956e

**Comment by mina86** [comment]
Is https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c7bf56d1a1d409c85e4fc514512bbd355 not an option?

**Comment by YizhePKU** [comment]
I think there should be a safe version of `as_chunks_unchecked`. 

Usually, an "unchecked" version of a function does the same thing as their checked counterpart. Currently, this is not the case.

I propose we rename the current `as_chunks_unchecked` to `as_chucks_exact_unchecked` and then add `as_chucks_exact`.

**Comment by c410-f3r** [comment]
What is the status of this feature?

**Comment by c410-f3r** [comment]
Looks like that checking of `N == 0 ` is the only blocker, right? If so, then we probably already have the answer from the responsible party.

* https://github.com/rust-lang/rust/pull/99471
* https://github.com/rust-lang/rust/pull/109049

In other words, it seems that officially runtime checks are preferable over compile-time checks so can we proceed with stabilization?

**Comment by jhpratt** [comment]
Permitting something to compile knowing with 100% certainty that it will panic seems like an odd decision to make, given that we have the capability to prevent it. Post-monomorphization errors aren't ideal, but surely it's better than an unconditional runtime panic?

**Comment by jDomantas** [comment]
The problem is that it's not "100% certainty that it will panic". There was an example in another thread (https://github.com/rust-lang/rust/pull/99471#issuecomment-1555908856) of code that is perfectly legal, but adding post-monomorphization error makes it fail to compile:

```rust
fn f<const N: usize>() {
    if N == 0 {
        ...
    } else {
        foo.as_chunks::<N>()...
    }
}

fn main() {
    f::<0>();
}
```

**Comment by jhpratt** [comment]
Ah, that makes sense. So unless we can solve the halting problem, a runtime error it is…

**Comment by mina86** [comment]
If the choice is between stabilising it as is with a runtime error, and keeping it experimental until `static if` and `where N > 0` becomes a thing, I vote for keeping this experimental.

**Comment by joboet** [comment]
I haven't found the time to work on it for the last few weeks, but I think my [pre-RFC for pattern types](https://gist.github.com/joboet/0cecbce925ee2ad1ee3e5520cec81e30#type-system) has a good solution for this when combined with https://github.com/rust-lang/project-const-generics/issues/26.

**Comment by jhpratt** [comment]
Now that [inline const is stabilized](https://github.com/rust-lang/rust/pull/104087), we should be able to make compilation fail when `N == 0`. It's not perfect (as mentioned in #99471, but it would be closer to the desired behavior. Likewise for `array_windows` (#75027)

**Comment by Xiretza** [comment]
I think the lang team has made it clear repeatedly that they prefer runtime panics over post-mono errors. Personally I don't agree, but I don't think it's worth arguing over anymore.

**Comment by jhpratt** [comment]
I don't agree either, but it's also not up to me. This issue has it crossed out, but that doesn't appear to be a formal decision in either direction. The docs currently say

> This check will most probably get changed to a compile time error before this method gets stabilized.

So it seems to remain an open question?

**Comment by c410-f3r** [comment]
A reduced example that should compile according to the responsible team.

```rust
pub fn foo<const N: usize>() -> i32 {
    const {
        assert!(N != 0);
    }
    1
}

pub fn bar<const N: usize>() -> i32 {
    if N == 0 {
        2
    } else {
        foo::<N>()
    }
}

fn main() {
    assert_eq!(bar::<0>(), 2);
}
```

Maybe something about the PR that unblocked inline constants (https://github.com/rust-lang/rust/pull/122568) could help here?

There are then 4 scenarios for stabilization:

1. Compile-time checks with the current CTFE behavior.
2. Compile-time checks with a machinery allows the above snippet.
3. https://github.com/rust-lang/rust/issues/74985#issuecomment-1766554603
4. Runtime checks, which I also personally disagree.

**Comment by jhpratt** [comment]
Thanks for that example. I've added it to the top comment.

**Comment by its-the-shrimp** [comment]
I haven't seen it discussed here yet, but another way to tackle the case of `N == 0` is to not tackle it and just make the iterator return `[T; 0]` until the end of time, what do you all think about this?

**Comment by ais523** [comment]
On the subject of the "what should happen if the const argument is 0" debate, I'd like to present what I think is an analogous situation:

```rust
fn divide<const D: u32>(n: u32) -> u32 {
    n / D
}

fn divide_checked<const D: u32>(n: u32) -> Option<u32> {
    if D == 0 {
        None
    } else {
        Some(n / D)
    }
}

fn main() {
    dbg!(divide_checked::<5>(99));
    dbg!(divide_checked::<0>(99));
    dbg!(divide::<5>(99));
    dbg!(divide::<0>(99));
}
```
[Playground.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=79a1a46d80f492484a857ab54e6c71e4)

The current behaviour of Rust in this situation is to compile with no errors or warnings (not even on the call to `divide::<0>`), and to panic at runtime. In other words, if the divisor is coming from a const generic, there's no check for validity, *even though* division by 0 is always meaningless, and even though `divide::<0>` always divides by 0 and can easily be observed at compile time to do so.

Meanwhile, if you just write `99/0` directly, you get a deny-by-default lint, `unconditional_panic`, which makes sense given that the code can never run successfully.

I don't think the situation of, e.g., `[T]::as_chunks::<0>` is significantly different from that of division by constant zero – nor do I think there would be any particular benefit from making them act differently. I also suspect that there is unlikely to be a better solution than "`[T]::as_chunks::<0>` panics at runtime if it is ever called, and produces a deny-by-default lint if you write the 0 explicitly, but not if the 0 is calculated based on a constant generic" – this gives the compile-time error we want in the cases where the statement is obviously meaningless, but allows the code to compile in cases where it might not be called due to expanded generic code.

It's also worth noting an edge case:
```rust
let fn_pointer = <[u32]>::as_chunks::<0>;
```
Do we want to allow this, or not? With the explicit `::<0>`, I can see a reasonable argument for throwing the `unconditional_panic` lint, with a message along the lines of "this function pointer will panic if it is ever called". I can't see a use for writing it directly (although I can see how it could potentially happen in expanded generic code, which could end up never calling the function in cases where the parameter is 0).

**Comment by Noratrieb** [comment]
The remaining issue here is what to do on `as_chunks::<0>()`. The current implementation panics, but with inline const it would also be possible to turn it into a post-mono error instead.

https://github.com/rust-lang/rust/issues/74985#issuecomment-2087680660 it is mentioned above that apparently the lang team prefers runtime panics over post-mono errors, but I don't see any links to discussion about this or justification for why.

The arguments for making it a post-mono error are clear: getting errors earlier (and at all if the code is not tested). Post mono errors won't error in `cargo check`, but will error in `cargo build`. Runtime panics won't error in `cargo check`, won't error in `cargo build` either, but will only (potentially) error in `cargo run`/`cargo test`/`cargo oops-its-prod-already`.

Other comments above make additional arguments in favor of post-mono errors, and the documentation of the methods also intended to make it a post-mono error eventually.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

I'm not sure whether what was mentioned about our lang preferences is actually true, so let's nominate to discuss.

**Comment by tgross35** [comment]
I don't remember what exactly the argument against post-mono errors was but I do remember thinking "that does makes sense", even though I prefer an earlier error. I want to say @BurntSushi may have originally expressed the concern?

If we do go with the runtime panic, we should document that a future Rust version is allowed to reject `N = 0` earlier. (It would be nice if this could be a `where N > 0` bound at some point in the future, but that would be more breakage than runtime->post mono).

**Comment by cuviper** [comment]
I think <https://github.com/rust-lang/rust/issues/74985#issuecomment-2293441106> is a good example why a runtime check is desirable, so users can be flexible in their own APIs around this. If I write `as_chunks::<N>()` where `N` might be zero *in unreachable code*, then I don't want an error.

**Comment by mina86** [comment]
That’s current limitation in expressiveness of the language.  I can
easily imagine:

```rust
impl [T] {
    fn as_chunks<const N: NonZeroUsize>() -> (&[[T; N]], &[T]);
}

fn foo<const N: usize>(slice: &[T]) -> Option<i32> {
    let chunks = const if let Some(N) = NonZeroUsize::new(N) {
        slice.as_chunks::<N>().0;
    } else {
        return None;
    };
    todo!()
}
```

That would require a few things from the language:
- Ability to take non-primitive types as const generic arguments.
- A ‘const if’ construct similar to C++’s ‘if constexpr’.
- An implicit conversion from integer literals to `NonZeroX` types.
  (Not strictly necessary but ergonomics would suffer greatly without
  it).

**Comment by traviscross** [comment]
Maybe neither here nor there, but note that the following works:

```rust
const fn divide<const D: u64>(n: u64) -> u64 {
    const { assert!(D > 0) }; // <--- Post-mono assertion.
    n / D
}

const fn divide_checked<const D: u64>(n: u64) -> Option<u64> {
    match D {
        0 => None,
        1.. => Some(divide::<D>(n)),
    }
}

fn main() {
    dbg!({const { divide_checked::<1>(1) }});
    dbg!({const { divide_checked::<0>(1) }}); // <-- Look here.
    dbg!(divide_checked::<1>(1));
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=%2F%2F+This+demonstrates%2C+based+on+what+we+allow+on+stable+Rust%2C+an%0A%2F%2F+observable+fact+about+how+things+work+with+respect+to%0A%2F%2F+monomorphization+and+consteval.%0A%2F%2F%0A%2F%2F+If+we+were+to+monomorphize+%60divide%3A%3A%3C0%3E%60%2C+we+might+witness%0A%2F%2F+post-mono+a+failure+of+its+const+assertion.%0A%2F%2F%0A%2F%2F+But%2C+today%2C+as+long+as+%60divide%60+is+generic+and+is+only+mentioned+in%0A%2F%2F+a+const+item+or+const+block%2C+then+we+accept+this+code.%0A%2F%2F%0A%2F%2F+There%27s+an+open+question+of+whether+we%27ll+commit+to+always%0A%2F%2F+accepting+this+code+that+we+have+not+yet+answered+definitively.%0A%2F%2F%0A%2F%2F+See%3A%0A%2F%2F%0A%2F%2F+-+https%3A%2F%2Fgithub.com%2Frust-lang%2Frust%2Fissues%2F122301%0A%2F%2F+-+https%3A%2F%2Fgithub.com%2Frust-lang%2Frfcs%2Fissues%2F3582+%28%60const+if%60%29%0A%2F%2F+-+https%3A%2F%2Fgithub.com%2Frust-lang%2Frust%2Fissues%2F74985%0A%2F%2F%0A%2F%2F+Author%3A+TC%0A%2F%2F+Date%3A+2025-02-23%0A%0Aconst+fn+divide%3Cconst+D%3A+u64%3E%28n%3A+u64%29+-%3E+u64+%7B%0A++++const+%7B+assert%21%28D+%3E+0%29+%7D%3B+%2F%2F+%3C---+Post-mono+assertion.%0A++++n+%2F+D%0A%7D%0A%0Aconst+fn+divide_checked%3Cconst+D%3A+u64%3E%28n%3A+u64%29+-%3E+Option%3Cu64%3E+%7B%0A++++match+D+%7B%0A++++++++0+%3D%3E+None%2C%0A++++++++1..+%3D%3E+Some%28divide%3A%3A%3CD%3E%28n%29%29%2C%0A++++%7D%0A%7D%0A%0Afn+main%28%29+%7B%0A++++dbg%21%28%7Bconst+%7B+divide_checked%3A%3A%3C1%3E%281%29+%7D%7D%29%3B%0A++++dbg%21%28%7Bconst+%7B+divide_checked%3A%3A%3C0%3E%281%29+%7D%7D%29%3B+%2F%2F+%3C--+Look+here.%0A++++dbg%21%28divide_checked%3A%3A%3C1%3E%281%29%29%3B%0A%7D%0A)

That is, if the function within which we want to compile-time assert is parametric over a type or const generic, is const, and we only call it from const items or within const blocks, then what we want here does work currently.

(The question of whether we'll guarantee to not later disallow this is #122301.)

Interestingly, this is the odd case of a call that we allow in a const context but disallow outside of one.

**Comment by the8472** [comment]
From the options in https://github.com/rust-lang/rust/issues/74985#issuecomment-2087722186 we would like a combination of point 2 and 3.

Runtime panics are not great because they fail _it compiles, it works_ or _make illegal states non-representatble_  principles, for precisely the thing that we're making legible to the type system.

The current CTFE approach, i.e. post-mono errors, is suboptimal because it would not be forwards-compatible with pattern types or where bounds. It would also make it difficult for a more-generic caller. The const-trick from #122301 ¹² would help with the latter issue but not with the former. It's also a somewhat obscure trick.

So we'd like `where N > 0` or `where N: usize is 1..` bounds on `array_chunks`. That's the first language feature we'd need. But on its own that would be insufficient for generic callers because they wouldn't have any way to generalize to all `N` via some fallback.

So the second desired feature is a way for generic callers to locally discharge the restriction. E.g. `const if`, inline `where` expressions or [unifying const generic impls](https://github.com/rust-lang/project-const-generics/issues/26)


```rust
// std function
fn chunks<const N: usize>()
where
    const { N > 0 },
{
    // ...
}

// user code
fn use_chunks<const N: usize>()  {
    where N > 0 {
       chunks<N>();
    } else {
       fallback();
    }
}
```

¹ That issue was actually motivated by this one, but it turned out insufficient.
² Additional prior discussion: https://github.com/rust-lang/rfcs/issues/3582#issuecomment-2140294966

**Comment by Noratrieb** [comment]
I agree with you that this would be a nice way to implement this. But I think this feature is so incredibly far off that we really shouldn't be blocking this very useful helper (that replaced chunks_exact, which is  lot more room for bugs) for literal decades because of one of its weird error edge cases.
Passing 0 is a mistake, and I don't think a lot of people will run into it. It's an obscure edge case, and I think both suggested ways of detecting it (post mono error or runtime error) are perfectly fine in practice, so I really think we should stabilize it with one of the two (whatever you prefer) instead of blocking it for decades.

**Comment by Noratrieb** [comment]
I would suggest going with a post mono error, as that has significant advantages. The only disadvantage is that it may impede flexibility. After stabilizing, in the post mono error, we could link to a tracking issue where people can complain if their use case was blocked by this inflexibility, and we should be able to make it a runtime error in case it's a problem (since that's backwards compatible as long as we document that people shouldn't rely on it being a post mono error).

**Comment by workingjubilee** [comment]
Would any of the people involved in the current state of the relevant const generic bounds involving pattern (types?)... I'm not entirely sure if I should be asking @oli-obk or @BoxyUwU... be able to weigh in on the projected development cycle for such a feature as libs-api is asking? @Noratrieb has projected "literal decades" and I have no reason to doubt such a guess, currently, but are we talking 1 or 10?

**Comment by oli-obk** [comment]
Uh, please do not put pattern types into the consideration for any API decisions. They are an *internal* language feature and will stay so for the forseeable future. Getting them anywhere production ready is definitely off far enough that waiting for them is at present like waiting for specialization to be stable. If you'd like to know more, join https://rust-lang.zulipchat.com/#narrow/channel/481660-t-lang.2Fpattern-types

**Comment by BoxyUwU** [comment]
I think there is basically no universe in which we get `N > 0` bounds in the next 2-3 years, it wouldn't surprise me if it took much longer than that even. Supporting `fn foo<const N: NonZeroU32>` is likely to be available much sooner (maybe the next year or so?), though the ability to take a generic `const N: usize` and pass that to a `const M: NonZeroU32` is once again *very* far off. rust-lang/project-const-generics#26 I also do not see happening for a long time, it would definitely be blocked on -Znext-solver being stable, and I don't know that anyone has really sat down and properly thought out how it should even be implemented (so lets also say definitely not happening in the next 2-3 years for this too).

One thing that could be interesting is adding an unstable attribute for const generics that requires the argument to be fully concrete, i.e. forbid `foo::<N>` and only allow `foo::<1>` or `foo::<{ MY_CONST }>`. This combined with PMEs would be fully forwards compatible (ish?) with whatever design we wind up with down the road for bounding const parameters (though it would rule out using `const N: NonZeroUsize` in the future). This wouldn't be as flexible on stabilization as desired but it seems more flexible than an error saying the function is unstable.

**Comment by the8472** [comment]
Ok then our options are

- runtime panic, no forward compatibility, more-general callers can use regular control flow
- postmono error as-is, no forward compatibility, more-general callers have to use #122301 forever
- wait for `const N: NonZeroUsize` to become usable. more-general callers would have to use #122301 but could (much) later upgrade if new language features that make it possible to construct one from a `where N > 0` or something like that
- request a lang feature to require concrete calls so we can maybe upgrade later, but it'll be much later too


> Uh, please do not put pattern types into the consideration for any API decisions.

I was gesturing at a general category of features to solve our problems by limiting the range of valid const values. If `NonZeroUsize` is easier than pattern types or where bounds that's useful information. If there are any other options then we'd like to hear them too.

**Comment by oli-obk** [comment]
I think just a mono time error seems fine. There are enough APIs that we would like to improve with better const checks, this one isn't particularly special. We can just ship it and handle it however we'll handle these in the future. Even a panic doesn't seem bad, we do that for lots of APIs in libstd

**Comment by the8472** [comment]
> There are enough APIs that we would like to improve with better const checks

Which stable ones are you thinking of? There are a bunch of array ones like this one that have been kept unstable due to this issue.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in lang triage today.  In particular, it had been asserted and then repeated in this thread that the lang team prefers runtime panics over post-mono errors.  That's not the case.  We like errors to happen as soon as possible.  So we prefer pre-mono errors over post-mono errors when possible, but similarly we prefer post-mono errors over runtime errors when that is possible and practical.

It's of course for libs-api to decide what's practical in this case.

Regarding what counsel we can provide on the language direction and timeline here, probably all of us want something in the spirit of `const if`, but as has been discussed on this thread, that or other similar solutions along this axis could take awhile.

**Comment by workingjubilee** [comment]
Sounds good.

It seems like our preferred paths are the slightly hacky `#[rustc_your_argument_must_be_this_concrete_to_pass]` or the monomorphizing error.

As the compiler magic that prevents jamming in random expressions is
- forward-compat with moving to a monomorphizing error
- forward-compat with moving to a panic
- NOT forward-compat with moving to a NonZeroUint type *unless*...
- ...with coercion from uint types to NonZeroUint, where it *is* forward-compat

It seems like the logical first move would be that we would not want NonZeroUint without something that could coerce values in the, er, nonzero range?

**Comment by crumblingstatue** [comment]
One of the truly magical things about `array_chunks` (and similar APIS, like `array_windows`) is that the length can be inferred, so you can write very ergonomic code like 
```Rust
for [a, b, c] in slice.array_chunks() {
    // do stuff with a, b, and c
}
```
If we want a solution that enforces a non zero uint, we should take care that it doesn't break inference.

**Comment by sffc** [comment]
Maybe just give chunk size 0 a well-defined behavior such as never evaluating the loop expression? 

And there can be a runtime panic in debug mode like for integer overflow.

I'd like to avoid a runtime panic being unconditionally compiled into code for N > 0. If we can assume that the compiler will eliminate it, maybe that's fine.

**Comment by GoldsteinE** [comment]
It’s an if on a const expression, so surely it will be optimized out.

**Comment by Qix-** [comment]
Just a thought, could array chunks simply return an Option and document that it returns None if the size is 0? It could even be extended such that if the windows size is 0 and the slice length is 0, that's well defined, returning Some.

Since the window size is constant, then in most cases it'd return `Some` unconditionally, which I assume would make any downstream users subject to constant folding optimizations. Also makes the behavior explicit, without needing any unstable syntax stuff.

**Comment by its-the-shrimp** [comment]
> Just a thought, could array chunks simply return an Option and document that it returns None if the window size is 0? It could even be extended such that if the windows size is 0 and the slice length is 0, that's well defined, returning Some.
> 
> Since the window size is constant, then in most cases it'd return `Some` unconditionally, which I assume would make any downstream users subject to constant folding optimizations. Also makes the behavior explicit, without needing any unstable syntax stuff.

This'd make the runtime value depend on the compile-time argument, which is not too good in terms of developer experience

**Comment by sffc** [comment]
> Maybe just give chunk size 0 a well-defined behavior such as never evaluating the loop expression?

Thought: with this approach, the loop expression becomes dead code, so it will generate a compiler warning.

**Comment by Kixunil** [comment]
@scottmcm good points!

All of the cases use iterators but that's survivorship bias - `chunks_exact` doesn't support random access so if we need random access anywhere we are not using the method there but something else. However I can't think of a good regexp that would find those places and I also can't think of where we would need it. Everything is simple enough to use iteration.

**Except splitting/subslicing of arrays.**

We have a bunch of places where we need to split *arrays, not slices* into sub-arrays of known length. So the operation is statically known to not fail in theory but not directly provable to the compiler. I wrote an extension trait for arrays that allows such splitting (also compile-time asserting that the length of the sub-arrays sums up to the original array) and even found a bug in our code. Pretty cool. So code like `let foo: &[u8; 8] = bar[0..8].try_into().expect("statically known len");` becomes `let foo = bar.split_array::<8, 8>();` if `bar: [u8; 16]` with the failed check causing post-mono error. I think this would be likely useful addition to `core` as well but more hairy than `as_chunks`/`array_chunks` because, as opposed to the `..` operator, this one takes two *lengths* of the arrays and also one of the lengths is in principle useless. So I think long-term `core` will want to have just one parameter and stabilizing a method with two parameters is not great.

Side note: I've implemented `as_chunks` rather than `array_chunks` so that once its iterated it has the unstable `TrustedLen` implemented to get access to optimizations. However now that I think about it I could've just added `chunk_arrays` (too) returning `core::slice::Iter<'_, [T; N]>`. But it does inform about #76354 a bit: in one case there was a check for remainder being empty which got much simpler (though could've been that simple with `chunks_exact` too) and in all other cases we were iterating within *array* where typing the extra `.0` was not too obnoxious (we were also passing it into `zip`, so no `.iter()` was needed. But it also reminds me we probably want to have something like `array_split_divisible` which compile-time-asserts that remainder is empty.

**Comment by fintelia** [comment]
In my code, the two versions that would be useful are `as_chunks` for cases where I want to process the remainder, and a different method that returned only the chucks but panicked if the slice length wasn't divisible by the chunk size. For instance, when working with RGBA data represented as a byte slice, you probably don't want to continuously add assertions about the slice length but it feels a bit ugly to use `array_chunks` and just silently ignore the the remainder.

**Comment by Kixunil** [comment]
@fintelia sounds like you may want to pass around `&[[u8; 4]]`, proving divisibility, instead of `&[u8]` so that you don't need to check (and panic) and only perform the check in the top-most function. I wouldn't be surprised if this also has additional performance benefits.

**Comment by fintelia** [comment]
@Kixunil yeah that could help for certain cases. Though [something like this part](https://github.com/image-rs/image-png/blob/master/src/filter.rs#L504-L579) of the png crate couldn't really benefit. The problem is that there's multiple different pixel formats supported, with some code generically handling them all as byte slices and other parts matching on each of the possible sizes and doing array operations on them.

**Comment by Kixunil** [comment]
@fintelia I'd use `enum Pixels<'a> { One(&'a [u8]), Two(&[[u8; 2]]), Three(&[[u8; 3]], ...)` instead of having `BytesPerPixel` and the pixels separated. I also see a lot of opportunity to reduce duplication in that code.

**Comment by Noratrieb** [comment]
This is getting really off topic, please stay on topic on tracking issues to keep them organized and avoid sending out too many notifications to people that are subscribed.

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting today. Overall we are happy with the current behavior using a runtime panic. Although a compile-time warning would be helpful, the scenario where it would be most useful likely involves generic wrappers forwarding `N` and we can only really warn at the immediate call site of `array_chunks` where `N` is not yet known.

@rfcbot merge

**Comment by Kixunil** [comment]
@Amanieu can someone actually show a production code that needs it to be panic rather than post mono error? Not hypothetical. So far I've only seen snippets saying `if N == 0 { what_the_hell_do_you_even_want_to_do_here()??? } else { do_something_senible() }` but none of them in real code.

I personally highly doubt this will ever be needed but that's not even that important: going from compile-time error to runtime error is not a breaking change but the reverse is. The more conservative approach (stabilizing with post-mono error) could've been taken for now and revisited in the future if there are people explaining how it limits them.

Frankly, I don't think Rust should add panicking APIs lightly. They lead people to get used to panics. I have a fairly recent cautionary tale. Because Rust had historically poor support for arrays (this got significantly better recently) a security-critical library I co-maintain had to operate on slices. We started changing that but we had a bunch of code in the form of `u32::from_le_bytes(foo[x..(x+4)].try_into().expect("len is 4"))` because there wasn't a better way in the standard library and until recently nobody realized it's possible to cause post-mono error. And because we had a bunch of code people got used to it to the point that **two** reviewers approved similar conversion in completely invalid context where the input comes from user. This could've lead to DoS vulnerability.

Thankfully, soon after that merged and before release I happened to realize that we can statically eliminate these panics by stuffing them into an extension trait that causes post-mono errors if used incorrectly. And that made me find that panic, along with another unrelated bug. (BTW we had a similar issue with `unsafe` creeping in that was resolved similarly. Thankfully we had no bugs there.)

To be clear, I'm not saying that runtime panic is certainly wrong *here*. If my assumption, that all code uses statically-known values, is true then all code will just hit those proposed warnings. But it does look very weird that a certain error is turned to runtime panic just to allow edge cases that could've had a separate `_rt_checked` method. And it also isn't great to enforce the "panicking is usual API design" sentiment.

**Comment by suprohub** [comment]
Yo finally

**Comment by Kixunil** [comment]
> But panicking on a const generic, in a way that's "well if you ever run the monomorphization you'll see it", is _way_ less of an issue.

That is certainly true in the vacuum. My story gives some perspective on what happens if it becomes a common pattern. The thing is, in that example all relevant values *were* hard-coded. If they were wrong they'd also always panic. But the issue is we got used to those conversions being the standard so much that a wrong conversion looked like the right one.

To map it to this method, one way this could go wrong is if people indeed started using `if N == 0` often they might get so used to it that one day they don't notice that the check is `if n == 0` and `n` is untrusted, so setting it to 0 is a DoS exploit. If used in a nested generic function it'd not emit any warning. And maybe for most, if not all, the people there's a more appropriate solution than doing that check but because the check is easy they weren't force to a more creative/principled solution. Just like we've completely missed that `get_array::<const LEN: usize>(offset: usize) -> Option<&[T; LEN]>` in an extension trait is possible and a good idea.

**Comment by RalfJung** [comment]
I thought the current policy was to always do FCP on the stabilization PR, not the tracking issue?

In this case, I think const-stabilizing these functions will require const-stabilizing an intrinsics (`exact_div`), which requires t-lang fcp.

**Comment by traviscross** [comment]
> I thought the current policy was to always do FCP on the stabilization PR, not the tracking issue?

It was reconsidered in the last libs-api meeting and decided that if that were to be changed, other related changes would need to be made to facilitate it, such as moving the discussions off of the tracking issues (or some other solution) so that the conversations would not be split between the tracking issue and the stabilization PR.  It was mentioned that libs-api had previously done the FCP on the stabilization PR and that the experience was considered to have been a failure, leading to the process of doing it on the tracking issue.  So the reasons behind the previous experience of failure with the process need to be explored and the causes resolved.

The divergence in experience is quite interesting, as on the lang side we've found doing the stabilization FCP on the stabilization PR to be a significant process improvement.

> In this case, I think const-stabilizing these functions will require const-stabilizing an intrinsic (`exact_div`), which requires t-lang fcp. Preparing the stabilization PR would uncover such points so we don't need to wait for a second round of FCP.

Agreed.  Similarly, apropos of @scottmcm's question, doing it on the stabilization PR helps to make the exact scope of the stabilization more clear.

**Comment by traviscross** [comment]
> In this case, I think const-stabilizing these functions will require const-stabilizing an intrinsic (`exact_div`), which requires t-lang fcp.

On this matter, of course, whoever is handling the stabilization, please of course put up a PR to stabilize the intrinsics and lang-nominate that.  We could make our FCP contingent on libs-api deciding to stabilize the functions here so the FCPs could run in parallel.

**Comment by RalfJung** [comment]
@rust-lang/libs-api it would be really good to get clarity on what exact set of functions and types is being proposed for stabilization here, as currently it is unclear.

**Comment by Amanieu** [comment]
> [@rust-lang/libs-api](https://github.com/orgs/rust-lang/teams/libs-api) it would be really good to get clarity on what exact set of functions and types is being proposed for stabilization here, as currently it is unclear.

It wasn't discussed in the detail in the meeting, we mostly focused on the `N == 0` issue. I'll bring it up again in tomorrow's meeting.

**Comment by Amanieu** [comment]
@rfcbot cancel

**Comment by Amanieu** [comment]
In the @rust-lang/libs-api meeting today we discussed the specific methods that we are considering stabilizing. We decided to only stabilize the `as_chunks` methods which return slices and leave the methods returning an iterator as unstable.

Methods being stabilized are:
```rust
impl [T] {
    const fn as_chunks<const N: usize>(&self) -> (&[[T; N]], &[T]);
    const fn as_rchunks<const N: usize>(&self) -> (&[T], &[[T; N]]);
    const unsafe fn as_chunks_unchecked<const N: usize>(&self) -> &[[T; N]];
    const fn as_chunks_mut<const N: usize>(&mut self) -> (&mut [[T; N]], &mut [T]);
    const fn as_rchunks_mut<const N: usize>(&mut self) -> (&mut [T], &mut [[T; N]]);
    const unsafe fn as_chunks_unchecked_mut<const N: usize>(&mut self) -> &mut [[T; N]];
}
```

@rfcbot merge

**Comment by BurntSushi** [comment]
The docs for these functions currently state:

> This check will most probably get changed to a compile time error before this method gets stabilized.

That should probably get fixed before stabilization or as part of stabilization.

**Comment by joboet** [comment]
> The docs for these functions currently state:
> 
> > This check will most probably get changed to a compile time error before this method gets stabilized.
> 
> That should probably get fixed before stabilization or as part of stabilization.

Couldn't we also remove the panic for these methods and return the entire slice as remainder if `N` is zero?

**Comment by cuviper** [comment]
> Couldn't we also remove the panic for these methods and return the entire slice as remainder if `N` is zero?

That's possible, but this statement in the docs would have to make an exception on the "strictly less" part:

> Splits the slice into a slice of `N`-element arrays, starting at the beginning of the slice, and a remainder slice with length strictly less than `N`.

**Comment by Kixunil** [comment]
I'd like to note that [my request](https://github.com/rust-lang/rust/issues/74985#issuecomment-2752017784) for explaining why take the less conservative approach and which real-world code needs to check `N` wasn't addressed yet. I don't find the responses so far satisfying because they do not include realistic code which needs the panic rather than compile error.

To put it differently, I don't have a problem with panics if some code actually needs them but I highly doubt there is any such production code. A single example would convince me that panic is better.

**Comment by joboet** [comment]
> But we could fix that by returning not `(&[], input)` but `(&[[]; usize::MAX], input)`!

I didn't think about the length of the slice being the (mathematically undefined) dividend – I guess my mind got these methods mixed up with `split_first_chunk`... please ignore my previous comment.

**Comment by joboet** [comment]
> A really interesting (but slightly off-topic) thing this makes me think about: [Okasaki](https://www.cambridge.org/core/books/purely-functional-data-structures/0409255DA1B48FA731859AC72E34D494) chapter 9 is about numerical representations, and the equivalence between a way of writing a number and a container of that length. The simplest example is that a linked list _is_ a [unary](https://en.wikipedia.org/wiki/Unary_numeral_system) representation of its own length. The book does a neat job of extending that to other things too, like what the container equivalent of a _binary_ number would be -- and how the `push`/`concat` implementation on the data structure ends up with the same complexity as the increment/addition operator on a binary number.
> 
> So from that perspective, `&[T]` _is_ a number, and `first_chunk` _is_ `checked_sub` on that number, `as_chunks` _is_ `/`+`%` on that number, and [`as_flattened`](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.as_flattened) _is_ `*`-by-a-constant on it.

This is getting very off-topic, but this equivalence is what I found so incredibly fascinating about lambda calculus. It's taught me that data is defined by how you interact with it – and these interactions are basically nothing but folds.

**Comment by shepmaster** [comment]
> I've read somewhere

Perhaps you mean [this comment](https://github.com/rust-lang/rust/issues/74985#issuecomment-2686269351)?

>  it had been asserted and then repeated in this thread that the lang team prefers runtime panics over post-mono errors. That's not the case. We like errors to happen as soon as possible. So we prefer pre-mono errors over post-mono errors when possible, but similarly we prefer post-mono errors over runtime errors when that is possible and practical

**Comment by Kixunil** [comment]
@shepmaster correct, thanks!

**Comment by Jules-Bertholet** [comment]
> I personally didn't respond because to me, examples aren't what convinces me to prefer panics. Post-mono errors are what convinces me to prefer panics. (Because I think post-mono errors should be avoided at almost all costs.)

Why? I’ve always thought of post-mono errors as being to pre-mono errors what runtime panics are to `Result::Err`. Yes, they are ugly, but there is a place for them. It’s always better to catch an error as soon as possible.

But, in this case, I would rather wait for the language features such that this could become a pre-mono error. The standard library is *forever*. If we stabilize a sub-par API now, we will be stuck with it for as long as people keep using Rust. There is no reason to rush; people can always use an external crate if they need this feature now.

**Comment by Kixunil** [comment]
FTR, I thought that post-mono errors can be changed to pre-mono but they can't:

```rust
fn foo<const N: usize>(x: &[u8]) {
    bar(x.as_chunks::<N>());
}

foo::<42>(y);
```

Would become invalid if `as_chunks` suddenly started requiring `foo` to add `where N != 0` bound.
So it'd be really great if it just forced people to pass in specific value.

**Comment by RalfJung** [comment]
>  So it'd be really great if it just forced people to pass in specific value.

That's probably possible for the compiler to enforce though it feels ad-hoc. Cc @BoxyUwU

**Comment by lcnr** [comment]
> That's probably possible for the compiler to enforce though it feels ad-hoc.

Very much possible, yeah 😁

**Comment by hanna-kruppe** [comment]
> I would rather wait for the language features (e.g., pattern types) such that this check could become a proper _pre-mono_ error. The standard library is _forever_. If we stabilize a sub-par API now, we will be stuck with it for as long as people keep using Rust. There is no reason to rush; people can always use an external crate if they need this feature now.

In practice what happens is not “people pull in some well-known crate that does it properly” (a quick search found one crate that provides it but it has exactly one dependent). Instead, they either contort their code to make do with stable standard library functions ([example I wrote myself recently](https://github.com/hanna-kruppe/chacha8rand/blob/3850414fb42a811e76960e0b44f401ab5a68553f/chacha8rand/src/array_ref.rs#L2)) that already exist, or they just implement it themselves with some `unsafe` (hopefully in a way that’s sound for N = 0). There is a small but nonzero amount of friction smeared across the Rust ecosystem due to the standard library not stabilizing this functionality. It’s not obvious to me that “the standard library is forever” outweighs that. The worst case outcome is that it either gets deprecated in a couple of years and a better way is added, or we’ll consider pre-mono errors better but not better enough to be worth the churn of deprecation. Either way, the cost is very small: it’s just a couple inherent methods that are marginally worse than they could’ve been in a perfect world, nothing with deep or far-reaching repercussions.

It’s not even obvious to me that there *will* be a clearly better way to have pre-mono errors in future Rust. There are vague ideas for what language features could enable them, but there’s no experience with how well they work in practice. So I don’t consider it a given that T-libs-api will say in five years “obviously the API should do XYZ to get a pre-mono error for N = 0”. Some way to put bounds on the values of const parameters will probably be needed eventually, but using it for preventing the specific error we’re talking about here may not be a good API design tradeoff after all.

**Comment by BoxyUwU** [comment]
@RalfJung yes I suggest something like that [upthread](https://github.com/rust-lang/rust/issues/74985#issuecomment-2684820652) but it seems like libs is not interested in deferring supporting generic callers until we know how things will pan out with enforcing the `N>0` requirement in the type system. as you noted it would prevent being able to generically wrap `array_chunks` callsites

**Comment by its-the-shrimp** [comment]
I agree with the sentiment to shrimply postpone this until there's a better way to enforce `N != 0`, this isn't an important trait, the absence of which creates discord in the ecosystem, it's just a couple utility methods, the current signature of which, shall it be stabilised now, will 100% be seen as an unfortunate, rushed decision a couple years down the road.

**Comment by its-the-shrimp** [comment]
And I don't think we'll _never_ get mechanisms for enforcing that at compile-time, e.g. changing `N: usize` to `N: NonZeroUsize` seems pretty much possible in the very near future, I wouldn't be pessimistic about this

**Comment by hanna-kruppe** [comment]
Note that even if `const N: NonZeroUsize` becomes possible next year (Boxy’s estimate upthread), it’ll be very unergonomic to use in some cases. For example, since there’s no NonZero literals, you’ll need to write something like `as_chunks::<{ NonZero::new(5) }>()` instead of just `as_chunks::<5>()`. Maybe hinting the length of the resulting arrays somewhere else could guide inference but I haven’t seen any discussion of how that would work. Similar problems apply to named constants that otherwise would be a plain usize.

**Comment by Sympatron** [comment]
If `as_chunks` would be stabilized with specific values only, it would be somewhat ad-hoc, but it would also be compatible with all other proposed solutions in the future without any breaking changes. And although it could not be used in all places where it might be useful, it could at least be used in many common cases and it could quite easily be stabilized in the near future.

This wouldn't be the first case where a subset of a feature is stabilized because it is useful enough on it's own (see const-generics).

**Comment by BurntSushi** [comment]
I largely agree with @hanna-kruppe here with regards to postponing. Also, just more broadly speaking, the downside here is not significant from my perspective. We're not talking about a subtle footgun that could lead to incorrect answers. We're talking about a slightly worse failure mode for one particular corner case. That on its own isn't enough to delay a useful API like this indefinitely IMO.

**Comment by Jules-Bertholet** [comment]
> In practice what happens is not “people pull in some well-known crate that does it properly” (a quick search found one crate that provides it but it has exactly one dependent). Instead, they either contort their code to make do with stable standard library functions (example I wrote myself recently) that already exist, or they just implement it themselves with some unsafe (hopefully in a way that’s sound for N = 0).

I agree that the present situation isn’t absolutely ideal. But I think the way to address it is something like Niko Matsakis’s [“preview crates” idea](https://smallcultfollowing.com/babysteps/blog/2025/01/29/preview-crates/), or the `futures` crate. Adding a dependency is far less of an issue if people know it’s blessed by the Rust team.

**Comment by hanna-kruppe** [comment]
In this particular case I don’t think that’ll really help. I’m not avoiding depending on a crate for this functionality because of trust, but because these functions are so trivial (ca. five lines of code each) that the complication and overhead of discovering a crate and depending on it is not worth it compared to the alternatives I described above.

A similar challenge applies on the Rust project side: who’s gonna be motivated to voluntarily take care of setting up a new repository and crate under the Rust project, publishing it, watch out for issues and PRs (unlikely in this specific case, but possible in theory), etc. when the code in question is so boring and the status quo is perfectly tolerable?

**Comment by BurntSushi** [comment]
Yup. Same. I would never depend on a micro-crate providing a function like this, no matter who built it.

**Comment by Amanieu** [comment]
We discussed this again the @rust-lang/libs-api meeting and concluded that there is little use for an `array_chunks` iterator when we already have `as_chunks`. As such we would like to remove this unstable API.

@rfcbot close

**Comment by scottmcm** [comment]
I'm pretty sure the bot can't do multiple FCPs in the same issue, even if the previous one has already ended.

I put a PR up (#143289) so you can do the FCP there if you'd like, @Amanieu

**Comment by crumblingstatue** [comment]
> there is little use for an `array_chunks` iterator when we already have `as_chunks`

`foo.as_chunks().0.iter().adapter()` is more convoluted than `foo.array_chunks().adapter()`, but I guess that alone might not warrant having it in the standard library.

**Comment by scottmcm** [comment]
> is more convoluted than [...]

I found this diff interesting: https://github.com/rust-lang/rust/pull/143289/files#diff-1a13885e89de65916e7e0f9436a79cb2e697e23dcaa1a24e8fb3106d8b89dc7eR905-R915

```diff
             (true, ([], v, [])) => Self::from_utf16_lossy(v),
             (true, ([], v, [_remainder])) => Self::from_utf16_lossy(v) + "\u{FFFD}",
             _ => {
-                let mut iter = v.array_chunks::<2>();
-                let string = char::decode_utf16(iter.by_ref().copied().map(u16::from_be_bytes))
+                let (chunks, remainder) = v.as_chunks::<2>();
+                let string = char::decode_utf16(chunks.iter().copied().map(u16::from_be_bytes))
                     .map(|r| r.unwrap_or(char::REPLACEMENT_CHARACTER))
                     .collect();
-                if iter.remainder().is_empty() { string } else { string + "\u{FFFD}" }
+                if remainder.is_empty() { string } else { string + "\u{FFFD}" }
             }
         }
     }
```

Notably in that while it did add an `.iter()`, it ended up *avoiding* a `.by_ref()`, and ended up about the same length overall.

So in context I suspect it'll often end up being not materially worse than before -- or if it is, it'll be in cases that were siliently ignoring the possibility of the `ArrayChunks::remainder`, which is itself a bit sketchy.

(And in places where ignoring the remainder was arguably fine because something else handled it, being forced to think about it might actually make that pre-handling more elegant, like seen in https://github.com/rust-lang/rust/pull/143289/files#r2178061170 )

**Comment by crumblingstatue** [comment]
> So in context I suspect it'll often end up being not materially worse than before -- or if it is, it'll be in cases that were siliently ignoring the possibility of the `ArrayChunks::remainder`, which is itself a bit sketchy.

I mostly use `array_chunks` in cases where the remainder (if any) can be safely ignored, but I understand that's probably not the majority use case. One example is iterating over RGBA color components.  It's very unlikely there are any remainder color components, and if there are, there is not much I can do about that, just ignore it. `array_chunks()` was more convenient and readable for this use case, but I understand that it's probably a niche use case.

**Comment by Kixunil** [comment]
@crumblingstatue I suspect this conversation doesn't belong here but I have no idea where to put it (feel free to suggest). In your case it sounds like a trailing value is actually *an error* and rather than ignoring it you should just return error. What you need is roughly:

```rust
fn get_rgba_data() -> Vec<u8> { /* obtains the data somehow */ }
fn validate_rgba(data: &[u8]) -> Result<&[[u8; 4]], InvalidLengthError> {
    let (rgba, tail) = data.as_chunks();
    if tail.is_empty() { Ok(rgba) } else { Err(InvalidLengthError(tail.len())) }
}
fn process_rgba(data: &[[u8; 4]]) {
    for &[r, g, b, a] in data { /* do whatever is needed here */ }
}

fn main() -> anyhow::Error {
    let rgba = get_rgba_data();
    let rgba = validate_rgba(rgba)?;
    process_rgba(rgba);
}
```

In short, use `&[[u8; 4]` as the newtype for *validated* RGBA slice, validate it at the program input and pass around only the validated one.

**Comment by scottmcm** [comment]
There's a nice version of that in the "remove `array_chunks`" PR above too:
```diff
-        if v.len() % 2 != 0 {
+        let (chunks, []) = v.as_chunks::<2>() else {
             return Err(FromUtf16Error(()));
-        }
+        };
```
So if there's no tail expected -- there's a bug somewhere if it happens -- then something like 
```rust
let (rgba, []) = data.as_chunks() else { unreachable!() };
```
might be a nice (and fuzzable!) way to communicate and catch that expectation.

(And yes, this is a fine place for this conversation because it's in context of whether there's sufficient value in the `array_chunks` iterator *in addition* to the already-stable stuff.)

**Comment by fintelia** [comment]
There's a case to be made that `as_chuncks_exact` should be added as shorthand for that. But I could also see not wanting to add all the combinations

**Comment by briansmith** [comment]
> let (rgba, []) = data.as_chunks() else { unreachable!() };

This doesn't work well with coverage reporting, currently: https://github.com/rust-lang/rust/issues/143149.

> There's a case to be made that `as_chuncks_exact` should be added as shorthand for that. 

This is what I do, having `as_chunks_exact` return an `Option<&[[T; N]]>` until the coverage stuff is fixed, so that I can use `.unwrap_or_else(|| unreachable!())` instead of `else { unreachable!() }`.

> But I could also see not wanting to add all the combinations

As well as maybe a long debate on whether it should return `Option<>` or panic.

**Comment by Kixunil** [comment]
`as_chunks_exact` would be nice to have if it returned `Result` where the error type contains the length of the trailing slice (preferably as `NonZeroUsize`). Because `Option` has the concise alternative @scottmcm suggested but if one wants to return `Result` with the information it requires one more line of code.
